### DISCUSSION TITLE
企业实践开源的动机 | 夜天之书

### DISCUSSION THREAD

**Discussion by giscus** [tisonkun/blog#25] [Comments]
# 企业实践开源的动机 | 夜天之书

随着开源软件全面占据软件供应链的各个阶段，商业公司开发基础软件或业务逻辑的时候，已经避不开对软件的使用了。经过一段时间对开源软件的使用，以及开源吞噬软件的趋势影响，研发能力突出的公司或团队，也会加入到开发开源软件的行列中来。 商业模型当中开源软件位置的不同，体现出企业实践开源动机的不同，并且会很大程度影响企业实践开源的行为。本文将讨论不同商业模型下，企业实践开源的动机和行为的差异。

https://www.tisonkun.org/2022/05/26/motivation-of-enterprise-open-source/

**Comment by manuzhang** [comment]
企业商业模式与开源软件之间的关系，个人以为还有这样一些角度：
* 开源软件的发展离不开一家商业公司在背后的持续投入开发与推广以及对社区的经营，比如早期的 DataStax 之于 Cassandra, Databricks 之于 Spark，现在 Presto / Iceberg / Hudi 的创始团队都从大公司出来创业，反面例子可能是 Storm
* 另一个是商业公司对开源软件与社区的「掌控」，比如哪些功能可以进入开源版本，成为 committer 的标准是什么，印象中 Databricks 在这一方面是很「严格」 的，以及「收编」了不少 committer（个人印象不一定对）。

**Reply by tisonkun** [reply]
> 开源软件的发展离不开一家商业公司在背后的持续投入开发与推广以及对社区的经营

有人发现了开源商业化的契机，进而依托开源软件创业，这是存在的。但是你说必须要有商业公司，否则开源软件无法发展，这是错的。从上世纪九十年代到本世纪初近二十年的时间，商业公司一贯地贬低和攻击开源软件，也没有阻止开源运动的存续和开源软件的发展。现在，商业公司发现了利用开源软件有利可图，于是开始用”控制“等等企业管理资本垄断的思路来”改造开源思想“。从我的角度看来，这与二十年前并无多大区别。

事实就是，没有商业公司，开源软件也可以很好的生存。商业公司推动开源软件在商业领域的落地，改善企业级开源软件使用的体验，是不可否认的。但是这里面没有强依赖关系。在这些商业公司提供相应的软件服务之前，诸如阿里巴巴、腾讯、推特、雅虎包括优步这样的公司，都有能力在商业环境下使用开源软件。

引用我之前回复另一位读者的邮件

> 开源运动过去二十多年都正常发展，资本前二十年肆意打压，近几年吹捧歪曲，都不会改变 Linux / PostgreSQL / Python / C++ / Kubernetes / Spring 这样的项目的发展轨迹。不存在什么不商业化开源就做不下去的杞人之忧。

关于”掌控“，这是一个值得一篇文章来讨论的话题。三两句话比较容易带偏方向。我认为社群之外的组织在社群当中的 Hard Influence 和 Soft Influence 是存在的，但是不应该简单地理解成”掌控“。大家容易看到 Databricks 在 Spark 社群的做法，和 Hadoop 三家相争在商业上的终局（实际上，Cloudera 也还活得好好的，只不过没有微软那么大而已）。但是还有更多的开源社群，比如 C++ 社群、Kubernetes 社群和 Apache BookKeeper 社群，这些都是多方参与且良性发展的开源社群，没人能够”掌控“它，但也不妨碍参与者从中获利。

**Reply by bestony** [reply]
关于一定要有商业公司这里，让我想起来了 FSF。这是一个非常典型的反例。🤣。

**Reply by manuzhang** [reply]
说明一下，我的原意不是「一定」，而是想补充一个这样的视角，比如在 Apache 项目中，有些发展得很成功，有些没能从 incubator 毕业，有些被「束之高阁」（attic），我们可以去探究下背后的原因。

**Reply by tisonkun** [reply]
@manuzhang 欢迎你做案例分析。按照是否毕业和有无商业公司参与，我可以想到不少例子，不过目前没时间去写。倒是有个“开源案例库”的企划，不过可能得接下来慢慢找时间填了。成功的项目都会有商业公司使用（未必是参与），这是因为作为软件会被开发者使用，进而进入到公司里。这就像 Ruby 语言生态的 Ruby on Rails 和 Chef 一样，大公司未必主观就要使用 Ruby，但是由于生态庞大优势明显，都会多少用上一点。这里面的因果关系和依赖关系，未必能够由简单的相关性推导出来。

也就是说，好的软件基本都会有商业公司使用，既然用了，其中研发力量较强的公司，为了保护自己的业务代码不被时代潮流抛弃，那么要么是 fork 二开，要么在此基础上还会和上游合作。这是我看到的当下和未来的趋势。这里不是因为商业公司用了，软件才活下来。但你说一个软件所有商业公司都不用，它的使用场景可能就缺少了这一块（还有很多生产力工具，未必在公司层面感知到使用了）。

目前一些议论把 judge 开源软件的权利隐约归纳到能不能赚钱，有没有商业公司青睐上，我觉得跟二十年前的傲慢没有太大的区别。我在多篇文章里表达的观点是，开源社群的目的是制造高质量的开源软件，由这种“石头汤”或者说分享精神共同造就出优秀的作品。这样的例子可以在 PostgreSQL / Apache Hudi / Apache BookKeeper 上面看到。关于这个角度，我可能在本月会出一份案例分析来讨论。

**Reply by tisonkun** [reply]
关于 attic 的事情，你可以看看最近 Apache Ambari 复活的事情，还挺有趣的。

Apache 讲究 Community over code 和 Independence 就在某种程度上防范社群凋零的主要根源。你可以看看我在[评注版《Apache 成熟度模型》](https://guides.tisonkun.org/apache-maturity-model/)里关于这两点的讨论。

<img width="773" alt="image" src="https://user-images.githubusercontent.com/18818196/173266409-3b5146fe-659d-4e7b-a76f-3226304d125d.png">

**Comment by FelixGibson** [comment]
谢谢作者写了这么好的文章。不过我有一个问题，我认为除了开源软件之外，还有一类开源产品，比如Bitwarden, Logseq, stable diffusion。它们开源的动机又是什么呢？会不会和开源软件有相同的地方也有差异？
比如Bitwarden虽然是一个开源密码管理工具，但是最近也拿到1亿美金的融资。
它开源的目的我认为有三个：
1. 开源是解决安全信任的最重要方式，这对密码管理工具来说很重要。
2. 开源+免费有助于积累声域和克服后发劣势，同时帮助 Bitwarden 带来反馈和推荐
3. Bitwarden的商业模型依托于开源软件。跟作者提到的第二点一样，开源版本有助于推广Bitwarden专门针对企业的企业版本，这也是Bitwarden收入的来源。
